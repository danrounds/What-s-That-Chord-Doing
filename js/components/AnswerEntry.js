import React from 'react';
import {connect} from 'react-redux';

import * as actions from  '../actions';

export class AnswerEntry extends React.Component {
    constructor(props) {
        super(props);
        this.onClick = this.onClick.bind(this);
        this.onKey = this.onKey.bind(this);

        this.keyMap = this.makeKeyMap();
        this.state = { guess: null };
    }

    componentWillReceiveProps(nextProps) {
        if (!nextProps.answeredCorrectly && nextProps.keyValue)
            this.onKey(nextProps);
        else if (!nextProps.guessN && this.state.guess !== null)
            this.setState({ guess: null });
    }

    onClick(e){
        if (!this.props.answeredCorrectly) {
            let guess;

            // We're either clicking a button or something deeper in our tree
            if (e.target.parentElement.nodeName === 'BUTTON')
                guess =  e.target.parentElement.innerText.split('\n')[0];
            else
                guess = e.target.innerText.split('\n')[0];

            this.setState({ guess });

            // actions
            if (guess === this.props.currentChord)
                this.props.dispatch(actions.markTurnCorrect());
            else
                this.props.dispatch(actions.incrementGuessN());
        }
    }

    onKey(props) {
        this.setState({ guess: this.keyMap[props.keyValue] || null });
        if (this.keyMap[props.keyValue]) {
            if (this.keyMap[props.keyValue] === props.currentChord)
                this.props.dispatch(actions.markTurnCorrect());
            else
                this.props.dispatch(actions.incrementGuessN());
        }
    }

    makeKeyMap() {
        const keys = ['a','s','d','f','g','h','j','k','l',';',
                      'z','x','c','v','b','n','m',',','.','/'];

        const keyMap = {};
        this.props.chordSubset.map((numeral) => {
            let k = keys.shift();
            keyMap[k] = numeral;
        });
        return keyMap;
    }

    makeButtons() {
        const keys = ['a','s','d','f','g','h','j','k','l',';',
                      'z','x','c','v','b','n','m',',','.','/'];


        return this.props.chordSubset.map((numeral) => {
            let k = keys.shift();

            return (
                <button className={this.state.guess === numeral
                                   ? (numeral === this.props.currentChord
                                      ? "buttonRightStyle"
                                      : "buttonWrongStyle"
                                     )
                        : "buttonStyle"}
                        key={numeral} onClick={this.onClick}>
                  {numeral}<br/>
                  {this.props.displayKeyboardShortcuts
                  && <div className="keyHint answerKeyHint">{k}</div>}
                </button>
            );
        });
    }

    render() {
        return (
            <div>
              {this.makeButtons()}
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    keyValue: state.game.keyValue,
    displayKeyboardShortcuts: state.game.displayKeyboardShortcuts,
    chordSubset: state.game.chordSubset,
    currentChord: state.game.chord,
    guessN: state.game.guessN,
    answeredCorrectly: state.game.answeredCorrectly
});

export default connect(mapStateToProps)(AnswerEntry);
