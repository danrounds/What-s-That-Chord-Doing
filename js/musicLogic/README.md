# Music logic

This is module that does all the "heavy lifting" to provide appropriate notes,
key-names, roman numerals, etc, to our music notation library, MIDI library,
and for our game logic.

The biggest things that it does are generates random chords and associated note 
data. It does this painstakingly, because the music notation library is utterly
literal about displaying enharmonic notes&#8212;that is, if we didn't carefully
manage these data, our display component would be completely screwed up.

The heart of the module is the `chordGetter` module, and its `init` and 
`getChord` methods. Everything else is subordinate methods or subordinate 
datatypes.

-DR, 2017
