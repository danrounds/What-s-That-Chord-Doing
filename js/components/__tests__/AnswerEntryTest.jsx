import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import {chordSets} from '../../musicLogic';
import {AnswerEntry} from '../AnswerEntry';

// AnswerEntry is one of our more "stateful" components. It deserves some tests
// with differing conditions

const props = {
    answeredCorrectly: false,
    displayKeyboardShorcuts: true,
    guessN: 4,
    chordSubset: chordSets.hardMinor,
    currentChord: 'vii°',
};

describe('Component: AnswerEntryTest', () => {
    it('should render', () => {
        const comp = renderer.create(<AnswerEntry {...props} />).toJSON();
        expect(comp).toMatchSnapshot();
    });

    it('should register wrong answers', () => {
        // Using the shallow renderer here, because it can .setState
        const comp = shallow(<AnswerEntry {...props} />);
        comp.setProps({ guess: 'vi' });
        expect(toJson(comp)).toMatchSnapshot();
    });

    it('should register right answers', () => {
        // Using the shallow renderer here, because it can .setState
        const comp = shallow(<AnswerEntry {...props} />);
        comp.setProps({ guess: 'vii°' }).render();
        expect(toJson(comp)).toMatchSnapshot();
    });

    it('should display keyboard shortcuts', () => {
        // We need `mount` to get the right level of depth to test whether or
        // not our keyHints are being displayed
        const comp = mount(<AnswerEntry displayKeyboardShortcuts={true} chordSubset={chordSets.hardMinor} />);
        expect(toJson(comp.find('.keyHint'))).not.toEqual(null);
    });

    it('should /not/ display keyboard shortcuts, when appropriate', () => {
        // We need `mount` to get the right level of depth to test whether or
        // not our keyHints are being displayed
        const comp = mount(<AnswerEntry displayKeyboardShortcuts={false} chordSubset={chordSets.hardMinor} />);
        expect(toJson(comp.find('.keyHint'))).toEqual(null);
    });

});
