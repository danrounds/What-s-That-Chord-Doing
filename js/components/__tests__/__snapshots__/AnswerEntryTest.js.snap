// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: AnswerEntryTest should register right answers 1`] = `
<div>
  <button
    className="buttonStyle"
    key="i"
    onClick={[Function]}
  >
    i
    <br />
  </button>
  <button
    className="buttonStyle"
    key="♭II"
    onClick={[Function]}
  >
    ♭II
    <br />
  </button>
  <button
    className="buttonStyle"
    key="ii°"
    onClick={[Function]}
  >
    ii°
    <br />
  </button>
  <button
    className="buttonStyle"
    key="ii"
    onClick={[Function]}
  >
    ii
    <br />
  </button>
  <button
    className="buttonStyle"
    key="♭III"
    onClick={[Function]}
  >
    ♭III
    <br />
  </button>
  <button
    className="buttonStyle"
    key="iv"
    onClick={[Function]}
  >
    iv
    <br />
  </button>
  <button
    className="buttonStyle"
    key="v"
    onClick={[Function]}
  >
    v
    <br />
  </button>
  <button
    className="buttonStyle"
    key="V"
    onClick={[Function]}
  >
    V
    <br />
  </button>
  <button
    className="buttonStyle"
    key="♭VI"
    onClick={[Function]}
  >
    ♭VI
    <br />
  </button>
  <button
    className="buttonStyle"
    key="vi"
    onClick={[Function]}
  >
    vi
    <br />
  </button>
  <button
    className="buttonStyle"
    key="♭VII"
    onClick={[Function]}
  >
    ♭VII
    <br />
  </button>
  <button
    className="buttonRightStyle"
    key="vii°"
    onClick={[Function]}
  >
    vii°
    <br />
  </button>
</div>
`;

exports[`Component: AnswerEntryTest should register wrong answers 1`] = `
<div>
  <button
    className="buttonStyle"
    key="i"
    onClick={[Function]}
  >
    i
    <br />
  </button>
  <button
    className="buttonStyle"
    key="♭II"
    onClick={[Function]}
  >
    ♭II
    <br />
  </button>
  <button
    className="buttonStyle"
    key="ii°"
    onClick={[Function]}
  >
    ii°
    <br />
  </button>
  <button
    className="buttonStyle"
    key="ii"
    onClick={[Function]}
  >
    ii
    <br />
  </button>
  <button
    className="buttonStyle"
    key="♭III"
    onClick={[Function]}
  >
    ♭III
    <br />
  </button>
  <button
    className="buttonStyle"
    key="iv"
    onClick={[Function]}
  >
    iv
    <br />
  </button>
  <button
    className="buttonStyle"
    key="v"
    onClick={[Function]}
  >
    v
    <br />
  </button>
  <button
    className="buttonStyle"
    key="V"
    onClick={[Function]}
  >
    V
    <br />
  </button>
  <button
    className="buttonStyle"
    key="♭VI"
    onClick={[Function]}
  >
    ♭VI
    <br />
  </button>
  <button
    className="buttonWrongStyle"
    key="vi"
    onClick={[Function]}
  >
    vi
    <br />
  </button>
  <button
    className="buttonStyle"
    key="♭VII"
    onClick={[Function]}
  >
    ♭VII
    <br />
  </button>
  <button
    className="buttonStyle"
    key="vii°"
    onClick={[Function]}
  >
    vii°
    <br />
  </button>
</div>
`;

exports[`Component: AnswerEntryTest should render 1`] = `
<div>
  <button
    className="buttonStyle"
    onClick={[Function]}
  >
    i
    <br />
  </button>
  <button
    className="buttonStyle"
    onClick={[Function]}
  >
    ♭II
    <br />
  </button>
  <button
    className="buttonStyle"
    onClick={[Function]}
  >
    ii°
    <br />
  </button>
  <button
    className="buttonStyle"
    onClick={[Function]}
  >
    ii
    <br />
  </button>
  <button
    className="buttonStyle"
    onClick={[Function]}
  >
    ♭III
    <br />
  </button>
  <button
    className="buttonStyle"
    onClick={[Function]}
  >
    iv
    <br />
  </button>
  <button
    className="buttonStyle"
    onClick={[Function]}
  >
    v
    <br />
  </button>
  <button
    className="buttonStyle"
    onClick={[Function]}
  >
    V
    <br />
  </button>
  <button
    className="buttonStyle"
    onClick={[Function]}
  >
    ♭VI
    <br />
  </button>
  <button
    className="buttonStyle"
    onClick={[Function]}
  >
    vi
    <br />
  </button>
  <button
    className="buttonStyle"
    onClick={[Function]}
  >
    ♭VII
    <br />
  </button>
  <button
    className="buttonStyle"
    onClick={[Function]}
  >
    vii°
    <br />
  </button>
</div>
`;
